### ISSUE TITLE
A multi-gpu training time issue

### ISSUE THREAD

**Issue by Zzzz2zy** [WongKinYiu/yolov9#283]
Great work. I have trained yolov9 using single gpu, 2 gpus, 4 gpus and 8 gpus respectively, but why the time per epoch remains the same? My script follows the readme.md, thanks for reply! 
```
python -m torch.distributed.launch --nproc_per_node 8 \
                                   --use_env \
                                   train_dual.py \
                                   --workers 8 \
                                   --device '0,1,2,3,4,5,6,7' \
                                   --sync-bn \
                                   --batch 128 \
                                   --data data/biaoji.yaml \
                                   --img 640 \
                                   --cfg models/detect/yolov9-c-bj.yaml \
                                   --weights './pretrained/yolov9-c.pt' \
                                   --name yolov9-c-bj \
                                   --hyp hyp.scratch-high.yaml \
                                   --min-items 0 \
                                   --epochs 500 \
                                   --close-mosaic 15
```

**Comment by Zzzz2zy** [comment]
> Do you slove it? I have same problem. :(

Yes I have. I set `--cache ram` and training time decreases when I use more gpus. You can try it.

**Comment by Wooho-Moon** [comment]
wonderful ! :) Thanks! umm.. Do you know the reason why the setting --cache ram decreases the training time?? Could you explain it for me?

**Comment by Zzzz2zy** [comment]
> Unfortunately, that solution does't work for me. My gpu machine whose size of ram is 256 G has 4 3090 gpus. But after setting --cache ram, All resource of rams are exhasuted, preventing the process or thread from reaching train loop. It is really wired because size of coco data set ( 2017 train ) is under 100G... It may copy some images in the ram.. But, I don't know exactly why this phenomia has occured. Could you give me some advise?

I'm sorry I don't figure out why this issue happens... In my case, the problem occurs only when I train the model on my own dataset, and all things go well when I train coco dataset. So perhaps we are facing different problems :( Maybe you can set `--cache disk` instead of ram? And make sure that all gpus are running.

**Comment by Wooho-Moon** [comment]
> > Unfortunately, that solution does't work for me. My gpu machine whose size of ram is 256 G has 4 3090 gpus. But after setting --cache ram, All resource of rams are exhasuted, preventing the process or thread from reaching train loop. It is really wired because size of coco data set ( 2017 train ) is under 100G... It may copy some images in the ram.. But, I don't know exactly why this phenomia has occured. Could you give me some advise?
> 
> I'm sorry I don't figure out why this issue happens... In my case, the problem occurs only when I train the model on my own dataset, and all things go well when I train coco dataset. So perhaps we are facing different problems :( Maybe you can set `--cache disk` instead of ram? And make sure that all gpus are running.

When I set --cache disk, there are no errors and unexpected happenings. Just one problem is that the time per epoch remains the same regardless of the number of GPUs used. Do you have any idea..?

**Comment by Wooho-Moon** [comment]
It seems like there might be a memory leak. The memory keeps increasing during training ( if I set --cache ram), indicating that allocations are being made somewhere in the code without corresponding deallocations. Debugging is likely needed to locate the specific part where memory is not being properly deallocated. Now, I'm trying to, but I don't figure out where the problem is yet.

**Comment by Zzzz2zy** [comment]
> > > Unfortunately, that solution does't work for me. My gpu machine whose size of ram is 256 G has 4 3090 gpus. But after setting --cache ram, All resource of rams are exhasuted, preventing the process or thread from reaching train loop. It is really wired because size of coco data set ( 2017 train ) is under 100G... It may copy some images in the ram.. But, I don't know exactly why this phenomia has occured. Could you give me some advise?
> > 
> > 
> > I'm sorry I don't figure out why this issue happens... In my case, the problem occurs only when I train the model on my own dataset, and all things go well when I train coco dataset. So perhaps we are facing different problems :( Maybe you can set `--cache disk` instead of ram? And make sure that all gpus are running.
> 
> When I set --cache disk, there are no errors and unexpected happenings. Just one problem is that the time per epoch remains the same regardless of the number of GPUs used. Do you have any idea..?

Weird... I tried on coco dataset but everything is fine, while on my own dataset the problem occurs. When I debug on my own dataset, I found out that the iteration time in each epoch is not consistent. Iteration time can be very short (about 5 seconds) and very long (over 1 minute), and it varies periodicly. Thus, I suppose that the dataloader waste lots of time and it's CPU and process of data IO that restrict the speed. And this seems to make sense because when you increase batch size, the burden on the CPU and IO process also increases accordingly, and that makes the training time remains the same. So I set cache to ram to reduce the time cost of CPU and IO, and it miraculously worked. But I'm still puzzled as to why this is happening, especially considering that everything went on well with the COCO dataset...
I hope this helps. 😀

**Comment by Wooho-Moon** [comment]
> teration time in each



> > > > Unfortunately, that solution does't work for me. My gpu machine whose size of ram is 256 G has 4 3090 gpus. But after setting --cache ram, All resource of rams are exhasuted, preventing the process or thread from reaching train loop. It is really wired because size of coco data set ( 2017 train ) is under 100G... It may copy some images in the ram.. But, I don't know exactly why this phenomia has occured. Could you give me some advise?
> > > 
> > > 
> > > I'm sorry I don't figure out why this issue happens... In my case, the problem occurs only when I train the model on my own dataset, and all things go well when I train coco dataset. So perhaps we are facing different problems :( Maybe you can set `--cache disk` instead of ram? And make sure that all gpus are running.
> > 
> > 
> > When I set --cache disk, there are no errors and unexpected happenings. Just one problem is that the time per epoch remains the same regardless of the number of GPUs used. Do you have any idea..?
> 
> Weird... I tried on coco dataset but everything is fine, while on my own dataset the problem occurs. When I debug on my own dataset, I found out that the iteration time in each epoch is not consistent. Iteration time can be very short (about 5 seconds) and very long (over 1 minute), and it varies periodicly. Thus, I suppose that the dataloader waste lots of time and it's CPU and process of data IO that restrict the speed. And this seems to make sense because when you increase batch size, the burden on the CPU and IO process also increases accordingly, and that makes the training time remains the same. So I set cache to ram to reduce the time cost of CPU and IO, and it miraculously worked. But I'm still puzzled as to why this is happening, especially considering that everything went on well with the COCO dataset... I hope this helps. 😀

Totally thanks for replying. I have a quick question. I found that when I use 4 gpus and I set --cache ram, coco datasets are copied 4times by train_loader during debugging. Is it right..? I don't know why same datasets are copied 4 times.

**Comment by Wooho-Moon** [comment]
> > > > Unfortunately, that solution does't work for me. My gpu machine whose size of ram is 256 G has 4 3090 gpus. But after setting --cache ram, All resource of rams are exhasuted, preventing the process or thread from reaching train loop. It is really wired because size of coco data set ( 2017 train ) is under 100G... It may copy some images in the ram.. But, I don't know exactly why this phenomia has occured. Could you give me some advise?
> > > 
> > > 
> > > I'm sorry I don't figure out why this issue happens... In my case, the problem occurs only when I train the model on my own dataset, and all things go well when I train coco dataset. So perhaps we are facing different problems :( Maybe you can set `--cache disk` instead of ram? And make sure that all gpus are running.
> > 
> > 
> > When I set --cache disk, there are no errors and unexpected happenings. Just one problem is that the time per epoch remains the same regardless of the number of GPUs used. Do you have any idea..?
> 
> Weird... I tried on coco dataset but everything is fine, while on my own dataset the problem occurs. When I debug on my own dataset, I found out that the iteration time in each epoch is not consistent. Iteration time can be very short (about 5 seconds) and very long (over 1 minute), and it varies periodicly. Thus, I suppose that the dataloader waste lots of time and it's CPU and process of data IO that restrict the speed. And this seems to make sense because when you increase batch size, the burden on the CPU and IO process also increases accordingly, and that makes the training time remains the same. So I set cache to ram to reduce the time cost of CPU and IO, and it miraculously worked. But I'm still puzzled as to why this is happening, especially considering that everything went on well with the COCO dataset... I hope this helps. 😀

I think Yolov9 is a awesome work. I read the paper and then I have a quick question.

1) Auxiliary branch is not used at inference stage, right? Then, when I export pt to onnx, Could I remove the Auxiliary branch? I have an example of onnx file. It seems like containing the auxiliary branch it may cause slowing down inference speed. Because onnxruntime execution provider may compute that brach. Would you offer some advice about above mentioned the approch of exporting onnx strategy?? I think it might be possibe since the pt file contains information about all weights of network...

**Comment by WongKinYiu** [comment]
You could use https://github.com/WongKinYiu/yolov9?tab=readme-ov-file#re-parameterization to remove aux branch.
